### PR TITLE
feat: support azure

### PR DESCRIPTION
**Pull request by jkoelker**
**What problem does this PR solve?**:

Adds Azure support to KiB

**Which issue(s) does this PR fix?**:

* https://jira.d2iq.com/browse/D2IQ-85255

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Initial Azure Support
```

### CODE REVIEW

**Comment by jkoelker** [comment]
I've build the release artifacts and uploaded them to: https://github.com/jkoelker/konvoy-image-builder/tree/jk/azure_snapshot/dist

Only Centos 7 is currently supported / tested. The easiest way to test this is to have the cli configured, but it also *should* support setting the following environment variables:

```bash
AZURE_LOCATION
AZURE_CLIENT_ID
AZURE_CLIENT_SECRET
AZURE_SUBSCRIPTION_ID
```

I believe there may be dragons with `EnsureImageDescriptions`, and may fail if the existing one already exists. I need to do more testing and make it idempotent. That said I'm not really thrilled with the idea of creating all the Gallery and ImageDescription, but azure images are a little funky in that every actual image is just a version of an existing description.

I'd like some feedback if we should even support creating them if they dont' exist, perhaps just checking that they do exist and failing is the right thing to do. I'm not really sure.

**Comment by jkoelker** [comment]
Welp looks like there is a bit more todo with the config merging.. setting back to draft for now.

**Comment by jkoelker** [comment]
I've build the release artifacts and uploaded them to: https://github.com/jkoelker/konvoy-image-builder/tree/jk/azure_snapshot/dist

Only Centos 7 is currently supported / tested. The follow env vars need to be set (everything except `AZURE_CLIENT_SECRET` can also be specified via CLI or override files):

```bash
AZURE_LOCATION
AZURE_CLIENT_ID
AZURE_CLIENT_SECRET
AZURE_TENANT_ID
AZURE_SUBSCRIPTION_ID
```

I'm not really thrilled with the idea of creating all the Gallery and ImageDescription, but azure images are a little funky in that every actual image is just a version of an existing description, and it requires the description to be created prior to packer running. 

I'd like some feedback if we should even support creating them if they don't exist, perhaps just checking that they do exist and failing is the right thing to do. I'm not really sure.

**Comment by jkoelker** [comment]
Rebased on main to remove last of dependent prs that are now merged.

File: cmd/konvoy-image-wrapper/cmd/wrapper.go [hunk 1]
```diff
@@ -166,6 +169,25 @@ func (r *Runner) setAWSEnv() error {
 	return nil
 }
 
+func (r *Runner) setAzureEnv() error {
```
Comment (supershal) [inline cmd/konvoy-image-wrapper/cmd/wrapper.go:172]: this function never returns any error. The return type should not be `error`

Comment (jkoelker) [reply cmd/konvoy-image-wrapper/cmd/wrapper.go:172]: Ah I forgot to have it get called in a loop, I'll remove the error return for now and if we ever another we can refactor it to have a common api for adding the env vars.

**Review by supershal — APPROVED** [review]

**Comment by jkoelker** [comment]
Just a rebase to get the changes from #251 to let the OVA offline tests to pass

**Comment by jkoelker** [comment]
Hrm, I'm not sure what the issue is with the 7.9 Offline OVA. It runs but then times out waiting for it to shutdown. I don't think anything in this PR is causing that (the release pr is also failing in the same way), and the 8.4 OVA does not fail. As I don't have vsphere access, @supershal @faiq if either of ya'll have time would you be able to look into this sometime?

If it is not an issue to just ignore it, let me know and I'll merge this in as is.

**Comment by jkoelker** [comment]
Rebased on main, and added the SSH args to ansible and packer sections in `pkg/packer/manifests/azure/packer.json.tmpl` to match aws template key acceptance.

**Comment by jkoelker** [comment]
My Azure trial subscription expired today, so I wasn't able to build it after rebasing, getting the subscription fixed and will build it. I'll hold off on merging until I get a validation run, since we dont' have CI for azure setup yet.

File: docs/cli/konvoy-image_build_azure.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      --gallery-image-sku string           the gallery image sku to set
+      --gallery-name string                the gallery name to publish the image in (default "konvoy")
+  -h, --help                               help for azure
+      --kubernetes-version string          The version of kubernetes to install. Example: 1.21.6
+      --location string                    the location in which to build the image (default "westus2")
+      --overrides strings                  a comma separated list of override YAML files
+      --packer-manifest string             provide the path to a custom packer manifest
+      --packer-on-error string             [advanced] set error strategy for packer. strategies [cleanup, abort, run-cleanup-provisioner]
+      --packer-path string                 the location of the packer binary (default "packer")
+      --resource-group string              the resource group to create the image in (default "konvoy")
```
Comment (joejulian) [inline docs/cli/konvoy-image_build_azure.md:35]: Should we be using "dkp" instead of "konvoy"?

File: pkg/azure/azure.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	galleriesClient := armcompute.NewGalleriesClient(subscriptionID, cred, nil)
+
+	pollerResp, err := galleriesClient.BeginCreateOrUpdate(
+		ctx,
+		description.ResourceGroupName,
+		description.GalleryName,
+		armcompute.Gallery{
+			Location: to.StringPtr(location),
+			Properties: &armcompute.GalleryProperties{
+				Description: to.StringPtr("This is gallery description."),
```
Comment (joejulian) [inline pkg/azure/azure.go:124]: Is this the correct string?

File: pkg/packer/manifests/azure/packer.json.tmpl [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    "image_version": "latest",
+    "konvoy_image_builder_version": "0.0.1",
+    "kubernetes_full_version": "",
+    "manifest_output": "manifest.json",
+    "plan_image_offer": "",
+    "plan_image_publisher": "",
+    "plan_image_sku": "",
+    "private_virtual_network_with_public_ip": "",
+    "resource_group_name": "konvoy",
+    "shared_image_gallery_name": "konvoy",
```
Comment (joejulian) [inline pkg/packer/manifests/azure/packer.json.tmpl:27]: Should these be `dkp`?

File: pkg/azure/azure.go (hunk 3 above)
Comment (jkoelker) [reply pkg/azure/azure.go:124]: Lulz, it is _a_ correct string ;). Updated to `DKP Gallery`.

**Comment by jkoelker** [comment]
Updated to switch `konvoy` -> `dkp` in the default created resource names, and rebased on main. Still working on getting my subscription updated to manually test.

**Pull request merged**